crates/hir_ty/src/diagnostics/decl_check.rs

```diff
@@ -888,6 +946,60 @@ fn main() {
         );
     }
 
+    #[test]
+    fn allow_attributes_crate_attr() {
+        check_diagnostics(
+            r#"
+    #![allow(non_snake_case)]
+
+    mod F {
+        fn CheckItWorksWithCrateAttr(BAD_NAME_HI: u8) {}
+    }
+
+    "#,
+        );
+    }
+
+    #[test]
+    #[ignore]
+    fn bug_trait_inside_fn() {
+        // FIXME:
+        // This is broken, and in fact, should not even be looked at by this
+        // lint in the first place. There's weird stuff going on in the
+        // collection phase.
+        // It's currently being brought in by:
+        // * validate_func on `a` recursing into modules
+        // * then it finds the trait and then the function while iterating
+        //   through modules
+        // * then validate_func is called on Dirty
+        // * ... which then proceeds to look at some unknown module taking no
+        //   attrs from either the impl or the fn a, and then finally to the root
+        //   module
+        //
+        // It should find the attribute on the trait, but it *doesn't even see
+        // the trait* as far as I can tell.
```

Comment:
Ah, this is probably because we don't really support function-local traits and impls yet

Reply:
I almost wonder if we should delete the part of this lint that recurses into functions to pick up items then, because we generate unavoidable warnings on bitfield macro output, that rustc ignores appropriately but we don't. It's unclear to me what a module is, in the terminology of the RA program representation, so I'm unsure what it's even trying to pick up by doing this recursion.

Reply:
I think it shouldn't be very hard to fix this, which we should do anyways (so we get IDE functionality in local impls). Though I haven't gotten a chance to look into that yet.